### ISSUE TITLE
Skipped tasks are not part of the BuildResult outcome

### ISSUE THREAD

**Issue by kasperfrederiksen** [gradle/gradle#2732]
Hi,

I am unsure if this is the intended behavior, but after upgrading to Gradle 4.1 we have this problem with our tests.

When doing a dry run, the BuildResult no longer include the skipped tasks.

Code example:

```
    def "testHelloWorldTask"(){
        given:
        buildFile << """

        task helloWorld {
            doLast {
                println 'Hello world!'
            }
        }
        """
        when:

        def result = GradleRunner.create()
                .withProjectDir(testProjectDir.getRoot())
                .withArguments("helloWorld", "--dry-run")
                .withPluginClasspath(pluginClasspath)
                .build();

        then:

        println(result.output)
        println("Tasks: $result.tasks")

        result.task(":helloWorld").getOutcome() == TaskOutcome.SKIPPED
    }

```
Output:

```
:helloWorld SKIPPED

BUILD SUCCESSFUL in 4s

Tasks: []

java.lang.NullPointerException: Cannot invoke method getOutcome() on null object
```
The list of tasks in buildResult is empty. If i remove the --dry-run parameter is works as i would expect.

Is this the expected behavior not to include "skipped" tasks in the org.gradle.testkit.runner.BuildResult?

Thanks for your help,
Kasper

**Comment by bmuschko** [comment]
Thanks for reporting. That looks like a regression to me. It works properly in 3.5. With 4.0 the task list is empty.

BTW: Why are you using `--dry-run` in a functional test. Don't you want to verify that specific task actions actually run?

**Comment by kasperfrederiksen** [comment]
No problem. 

The example is a bit more complicated than the above code. We use it to include some tags that are used to generate asciidoc documentation. The test will fail if the parameters for our custom plugin has changed. This way the documentation is always up to date. Therefore only dry-run is needed. We have other tests that tests the actually implementation.

**Comment by blindpirate** [comment]
After some investigation, it seems that build start and end event are not triggered in [DryRunBuildExecutionAction](https://github.com/gradle/gradle/blob/master/subprojects/core/src/main/java/org/gradle/execution/DryRunBuildExecutionAction.java#L37). With `--dry-run`, the subsequent [SelectedTaskExecutionAction](https://github.com/gradle/gradle/blob/master/subprojects/core/src/main/java/org/gradle/execution/SelectedTaskExecutionAction.java#L37) won't be executed, [where](https://github.com/gradle/gradle/blob/master/subprojects/core/src/main/java/org/gradle/execution/taskgraph/DefaultTaskGraphExecuter.java#L230) start and end events are triggered.

It is introduced in [https://github.com/gradle/gradle/commit/aa469f68072d0b19a6f94202d3c180eaf16a6201](https://github.com/gradle/gradle/commit/aa469f68072d0b19a6f94202d3c180eaf16a6201), @lacasseio What if I restore the `--dry-run` behavior to `task.setEnabled(false)`?

**Comment by bmuschko** [comment]
I talked to @lacasseio about this before. The decision to not resolve the task graph has been made deliberately. Nevertheless, I still think that the use case of checking the Gradle tasks of a build from a build execution with the Tooling API is a valid one.

**Comment by bmuschko** [comment]
> We use it to include some tags that are used to generate asciidoc documentation. The test will fail if the parameters for our custom plugin has changed. This way the documentation is always up to date. Therefore only dry-run is needed. 

I still don't understand that use case. Maybe you can put together an example for clarity.

**Comment by oehme** [comment]
@lacasseio could we just switch printing task headers on when `--dry-run` is passed? Then the actual dry-run logic could return to the previous behavior.

**Comment by grv87** [comment]
Use case is testing task dependencies as described [here](http://mrhaki.blogspot.ru/2014/11/gradle-goodness-check-task-dependencies.html)

**Comment by bmuschko** [comment]
@grv87 IMHO while you can of course execute a build with `dry-run` via TestKit I still don't think this is really a use case. When writing test cases you actually want to execute task actions and what to make assertions based on that.

**Comment by grv87** [comment]
@bmuschko:
1. Tasks are provided by other plugins. I believe it is bad practice to test third-party code.
2. Actual execution of tasks requires much more configuration and so brings more complexity and fragility.
3. Tasks could have side-effects such as publishing artifacts to repositories.

By specification plugin under test should configure task dependencies - and the test should confirm that only, nothing else.

**Comment by bmuschko** [comment]
@grv87 

1. If your plugin applies other plugins then you should also test the integration by executing the tasks that you would usually expect to be executed and verify if it works with your plugin.
2. You are actually not verifying much by just inspecting the task execution order. There are plenty of other cases where your plugin still is not functioning properly (e.g. up-to-date checking, things that should be created so it can be used by your plugin to work properly).
3. If you expect a task to publish then you should set up your test case to publish to a file-local repository to inspect the result.

**Comment by ajoberstar** [comment]
@bmuschko In some cases the order _is_ the test though. For example, if you have a plugin the defines ordering rules (via shouldRunAfter, mustRunAfter, etc) a test with dry run would be the most efficient way to verify that behavior.

**Comment by stewartbryson** [comment]
> @bmuschko In some cases the order _is_ the behavior though. For example, if you have a plugin the defines ordering rules (via shouldRunAfter, mustRunAfter, etc) a test with dry run would be the most efficient way to verify that behavior.

This is exactly my use case @ajoberstar. I simply want to make sure that a particular lifecycle task does  execute all the dependencies I expect, in the order I expect.

**Comment by stewartbryson** [comment]
> @grv87 IMHO while you can of course execute a build with `dry-run` via TestKit I still don't think this is really a use case. When writing test cases you actually want to execute task actions and what to make assertions based on that.

Certainly... I want some of my tests to test that certain actions actually occur. But sometimes I simply want to make sure that a particular lifecycle task does execute all the dependencies I expect, in the order I expect. With a --dry-run, I can pass a bunch of different properties and task inputs using Spock data-driven testing, and simply test that the graph is as I expect. Then, I have further tests that will test individual behavior, and some of that might involve integration testing.

**Comment by stewartbryson** [comment]
FYI... I use the following to get a list of tasks in order from the output of a --dry-run execution:

```
List tasks = result.output.readLines().grep(~/(:)(.+)/).collect { 
   it.replaceAll(/(:)(\w+)( SKIPPED)*/, '$2')
}
```

**Comment by grv87** [comment]
Checking list of tasks run does not really test that task ordering works. It is a flaky test.
List of tasks is always in some order, but this order may be random.
It could happen that this random order matches required.

**Comment by grv87** [comment]
Actually, @blindpirate already mentioned workaround.
Disable all tasks and and run Gradle in standard way:
```
    def "testHelloWorldTask"(){
        given:
        buildFile << """

        task helloWorld {
            doLast {
                println 'Hello world!'
            }
        }

        tasks.all {
            enabled = false
        }
        """
        when:

        def result = GradleRunner.create()
                .withProjectDir(testProjectDir.getRoot())
                .withArguments("helloWorld")
                .withPluginClasspath(pluginClasspath)
                .build();

        then:

        println(result.output)
        println("Tasks: $result.tasks")

        result.task(":helloWorld").getOutcome() == TaskOutcome.SKIPPED
    }
```

**Comment by ov7a** [comment]
Sorry for the late reply.

The issue is in the backlog of the relevant team and is prioritized by them.

***

As described, there's a [workaround](https://github.com/gradle/gradle/issues/2732#issuecomment-440725747).

**Comment by jbartok** [comment]
Just a comment for whomever will look at this. What makes this clearly an issue is the inconsistency between using `--dry-run`  and what the workaround does (`tasks.all { enabled = false }`). These two things should work exactly the same way, but they don't, hence it's inconsistent/bug.
